packages/compass-collection/src/modules/collection-tab.ts

```diff
@@ -0,0 +1,244 @@
+
+export const selectTab = (
+  tabName: CollectionState['currentTab']
+): CollectionThunkAction<void> => {
+  return (dispatch, _getState, { localAppRegistry }) => {
+    dispatch({ type: CollectionActions.ChangeTab, tabName });
+    localAppRegistry.emit('subtab-changed', tabName);
+  };
+};
+
+export const selectDatabase = (): CollectionThunkAction<void> => {
+  return (dispatch, getState, { globalAppRegistry }) => {
+    const { metadata } = getState();
+    const { database } = toNs(metadata.namespace);
+    globalAppRegistry.emit('select-database', database);
+  };
+};
+
+export const editView = (): CollectionThunkAction<void> => {
+  return (dispatch, getState, { globalAppRegistry }) => {
+    const { metadata } = getState();
+    globalAppRegistry.emit('collection-tab-modify-view', {
+      ns: metadata.namespace,
+    });
+  };
+};
+
+export const returnToView = (): CollectionThunkAction<void> => {
+  return (dispatch, getState, { globalAppRegistry }) => {
+    const { editViewName } = getState();
+    globalAppRegistry.emit('collection-tab-select-collection', {
+      ns: editViewName,
+    });
+  };
+};
+
+const setupRole = (
+  roleName: string
+): CollectionThunkAction<{ name: string; component: React.ReactElement }[]> => {
+  return (
+    dispatch,
+    getState,
+    { localAppRegistry, globalAppRegistry, dataService }
+  ) => {
+    const roles = globalAppRegistry.getRole(roleName) ?? [];
+
+    return roles.map((role) => {
+      localAppRegistry.registerRole(roleName, role);
+
+      const {
+        name,
+        component,
+        storeName,
+        configureStore,
+        actionName,
+        configureActions,
+      } = role;
+
+      const collectionStoreMetadata = {
+        ...getState().metadata,
+        localAppRegistry,
+        globalAppRegistry,
+        dataProvider: {
+          // Even though this is technically impossible, all scoped plugins
+          // expect error key to be present
+          error: null,
+          dataProvider: dataService,
+        },
+        query: getState().initialQuery,
+        aggregation: getState().initialAggregation,
+        editViewName: getState().editViewName,
+      };
+
+      let actions;
+      if (actionName && configureActions) {
+        actions =
+          localAppRegistry.getAction(actionName) ??
+          localAppRegistry
+            .registerAction(actionName, configureActions())
+            .getAction(actionName);
+      }
+
+      let store;
+      if (storeName && configureStore) {
+        store =
+          localAppRegistry.getStore(storeName) ??
+          localAppRegistry
+            .registerStore(
+              storeName,
+              configureStore({ ...collectionStoreMetadata, actions })
+            )
+            .getStore(storeName);
+      }
+
+      return {
+        name,
+        component: React.createElement(component, { store, actions }),
+      };
+    });
+  };
+};
+
+export const renderScopedModals = (): CollectionThunkAction<
```

Comment:
This might look a bit weird, but allows us to separate `appRegistry` from the rendering logic in a clean way: actions have access to services (like registry) through thunk extra arg and components use a pretty common "render prop" interface, so both store and components are using  conventional logic to handle rendering plugins from registry.

This should also go away and get encapsulated with some additional work that we are planning in the scope of "compass web", hence the TODO below